src/analyzer/protocol/ldap/ldap.spicy

```diff
@@ -130,29 +130,100 @@ public type Result = unit {
 };
 
 type Ctx = struct {
-  saslStripping: SaslStripping;  # Which mode of SASL stripping to use.
+  messageMode: MessageMode;   # Message dispatching mode
+  saslMechanism: string;      # The SASL mechanism selected by the client.
+  startTlsRequested: bool;    # Did the client use the StartTLS extended request?
 };
 
 #-----------------------------------------------------------------------------
 public type Messages = unit {
   %context = Ctx;
-  : SASLStrip(self.context())[];
+  : MessageDispatch(self.context())[];
 };
 
 #-----------------------------------------------------------------------------
-public type SASLStrip = unit(ctx: Ctx&) {
-  switch( ctx.saslStripping ) {
-    SaslStripping::Undef -> : Message(ctx);
-    SaslStripping::MS_KRB5 -> : SaslMsKrb5Stripper(ctx);
+public type MessageDispatch = unit(ctx: Ctx&) {
+  switch( ctx.messageMode ) {
+    MessageMode::Undef -> : Message(ctx);
+    MessageMode::MS_KRB5 -> : SaslMsKrb5Stripper(ctx);
+    MessageMode::TLS -> : TlsForward;
+    MessageMode::MAYBE_ENCRYPTED -> : MaybeEncrypted(ctx);
+    MessageMode::CLEARTEXT -> : Message(ctx);
+    MessageMode::ENCRYPTED -> : EncryptedMessage;
   };
 };
 
 
+#-----------------------------------------------------------------------------
+type MaybeEncrypted = unit(ctx: Ctx&) {
+  # A plaintext LDAP message always starts with at least 3 bytes and the first
+  # byte is 0x30 for the sequence. A SASL encrypted message starts with a 4 byte
+  # length field. The heuristic here is that if we see 0x30, assume it's
+  # unencrypted. This should be pretty good, as we deal with a LDAP message
+  # of minimally size 0x30000000 (~768 MB) and that seems unlikely.
+  var start: iterator<stream>;
+  var saslLen: uint64;
+  var mech: bytes;
+
+  on %init {
+    self.start = self.input();
+    # Don't have starts_with() on string, work around that.
+    self.mech = ctx.saslMechanism.encode(spicy::Charset::UTF8);
+  }
+
+  first: uint8 {
+    if ( $$ == 0x30 ) {
+      ctx.messageMode = MessageMode::CLEARTEXT;
+    } else {
+      ctx.messageMode = MessageMode::ENCRYPTED;
+    }
+  }
+
+  # As a further heuristic, if encrypted mode was decided and the client
+  # requested GSSAPI or GSS-SPNEGO (or we just didn't see it) peak a bit
+  # into the SASL payload and check if it starts with a 0504 (WRAP_TOKEN).
+  # If so, switch into KRB mode assuming that's what is being used and
+  # have a chance seeing some more plaintext LDAP in non-sealed tokens.
+  rem: uint8[3] if ( ctx.messageMode == MessageMode::ENCRYPTED && (|self.mech| == 0 || self.mech.starts_with(b"GSS")) ) {
+    self.saslLen = (self.first << 24) + ($$[0] << 16) + ($$[1] << 8) + $$[2];
+  }
+
+  : uint16 if ( self.saslLen >= 2 ) {
+    if ( $$ == 0x0504 ) {
+      ctx.messageMode = MessageMode::MS_KRB5;
+    }
+  }
+
+  # Rewind the input.
+  : void {
+    self.set_input(self.start);
+  }
+
+  # One recursion to parse with the new ctx.messageMode setting.
+  : MessageDispatch(ctx);
```

Comment:
Nit: the dispatch feels slightly scary as this unit consumes no data and in some unfortunate future could recurse indefinitely. Might be worth adding an `assert ctx.messageMode != MessageMode::MAYBE_ENCRYPTED` to the rewinding hook.

Also `MaybeEncrypted` will not expose its payload since the `MessageDispatch` is an anon field. Would it make sense to name it, or maybe even `&convert=self.dispatched` when `MaybeEncrypted` is done parsing to transparently remove the indirection?

Reply:
I'm adding the `assert`.

I'm not following the consequences of keeping MessageDispatch anonymous. The `Message` within `MessageDispatch` still generates events properly I believe.

Reply:
> I'm not following the consequences of keeping MessageDispatch anonymous. The Message within MessageDispatch still generates events properly I believe.

Yes, this does not affect events which are triggered via units down the parse tree, but is more about what "interface"  `MaybeEncrypted` exposes; its current shape completely dictated by parsing. Since `MessageDispatch` contains useful data it might make sense to expose it to users, but we currently have none 🤷. Feel free to drop.